src/lua/serializer.c

```diff
@@ -433,6 +552,129 @@ lua_field_tostring(struct lua_State *L, struct luaL_serializer *cfg, int idx,
 	return luaL_tofield(L, cfg, idx, field);
 }
 
+/**
+ * Get the next field from the table in unsorted order.
+ * @param L Lua stack.
+ * @param cfg Serializer configuration.
+ * @retval 0 if there are no more fields in the table.
+ */
+static int
+luaL_next_field_unsorted(struct lua_State *L)
+{
+	assert(lua_istable(L, -3));
+
+	/* const table, table, nil, key */
+	int r = lua_next(L, -3);
+
+	if (r == 0) {
+		/**
+		 * No more elements in the table. Delete the copy of the table.
+		 * const table, table, nil
+		 */
+		lua_pop(L, 2);
+		/* const table */
+		return 0;
+	}
+
+	return r;
+}
+
+int
+luaL_next_field(struct lua_State *L, struct luaL_serializer *cfg)
+{
+	/* At least: const table, key_idx, key */
+	assert(lua_gettop(L) >= 3);
+	assert(lua_istable(L, -3));
+
+	/**
+	 * Both key_idx and key are nil, first call, copy the table
+	 * so that we can remove printed elements from its copy, leaving
+	 * the initial table intact.
+	 */
+	if (lua_isnil(L, -1) && lua_isnil(L, -2)) {
+		/* const table, nil, nil */
+		lua_pop(L, 2);
+		lua_newtable(L);
+		/* const table, table */
+
+		/* Copy the table. */
+		lua_pushnil(L);
+		/* const table, table, nil */
+		while (lua_next(L, -3)) {
+			/* const table, table, key, value */
+			lua_pushvalue(L, -2);
+			/* const table, table, key, value, key */
+			lua_pushvalue(L, -2);
+			/* const table, table, key, value, key, value */
+			lua_settable(L, -5);
+			/* const table, table, key, value */
+			lua_pop(L, 1);
+			/* const table, table, key */
+		}
+		/* const table, table */
+		lua_pushnil(L);
+		lua_pushnil(L);
+		/* const table, table, nil, nil */
+	}
```

Comment:
Probably, I'm wrong, but there's no need to do a full copy of the table if the encode key order isn't specified and the method `luaL_next_field_unsorted()` doesn't delete the table copy.

I think extra table deletion can be done after going through the whole unsorted part.

Making an unnecessary full copy is quite expensive.

Reply:
You are right that this might be unnecessary, but then the lau stack will be different for different options of the call. I've tried to avoid it, because I didn't want to deal with it) will look more into it too.

Reply:
Ok. I've looked into this a bit more, and don't see an easy way to fix it, without complicating the logic and adding more functions similar to `luaL_next_field_unsorted()`.

I use `luaL_next_field_unsorted()` in two conditions:

1. There is no sorting order specified at all.
2. We have finished going through sorted part of the array and only need to go through what is left in arbitrary order.

The second branch obviously needs the copied array, from which we removed elements during iterating over the sorted part.

The first branch I was able to optimize only by introducing a new function, similar to `luaL_next_field_unsorted()`. If you or @Totktonada sees any better solution - feel free to propose, I would gladly improve this part.

<details>

<summary>
Full diff
</summary>

diff --git a/src/lua/serializer.c b/src/lua/serializer.c
index a892d067a5..424c807701 100644
--- a/src/lua/serializer.c
+++ b/src/lua/serializer.c
@@ -551,14 +551,17 @@ lua_field_tostring(struct lua_State *L, struct luaL_serializer *cfg, int idx,
 }

 /**
- * Get the next field from the table in unsorted order.
+ * Get the next field from the leftovers table in unsorted order.
  * @param L Lua stack.
  * @retval 0 if there are no more fields in the table.
  */
 static int
 luaL_next_field_unsorted(struct lua_State *L)
 {
+       assert(lua_gettop(L) >= 4);
+       assert(lua_isnil(L, -2));
        assert(lua_istable(L, -3));
+       assert(lua_istable(L, -4));

        /* const table, table, nil, key */
        int r = lua_next(L, -3);
@@ -576,13 +579,47 @@ luaL_next_field_unsorted(struct lua_State *L)
        return r;
 }

+/**
+ * Get the next field from the initial table in unsorted order.
+ * @param L Lua stack.
+ * @retval 0 if there are no more fields in the table.
+ */
+static int
+luaL_next_field_fallback(struct lua_State *L)
+{
+       assert(lua_gettop(L) >= 3);
+       assert(lua_isnil(L, -2));
+       assert(lua_istable(L, -3));
+
+       /* const table, nil, key */
+       int r = lua_next(L, -3);
+
+       if (r == 0) {
+               /*
+                * const table, nil
+                */
+               lua_pop(L, 1);
+               /* const table */
+               return 0;
+       }
+
+       /* const table, nil, key, value */
+       return r;
+}
+
 int
 luaL_next_field(struct lua_State *L, struct luaL_serializer *cfg)
 {
-       /* At least: const table, key_idx, key */
+       /* const table, (table ?), key_idx, key */
        assert(lua_gettop(L) >= 3);
+       assert(lua_isnil(L, -2) || lua_isnumber(L, -2));
        assert(lua_istable(L, -3));

+       /* No need to sort anything */
+       if (!cfg->encode_key_order) {
+               return luaL_next_field_fallback(L);
+       }
+
        /*
         * Both key_idx and key are nil, first call, copy the table
         * so that we can remove printed elements from its copy, leaving
@@ -614,11 +651,6 @@ luaL_next_field(struct lua_State *L, struct luaL_serializer *cfg)
                /* const table, table, nil, nil */
        }

-       /* No need to sort anything */
-       if (!cfg->encode_key_order) {
-               return luaL_next_field_unsorted(L);
-       }
-
        /* Going through the unsorted part */
        if (!lua_isnil(L, -1) && lua_isnil(L, -2)) {
                return luaL_next_field_unsorted(L);
diff --git a/third_party/lua-cjson/lua_cjson.c b/third_party/lua-cjson/lua_cjson.c
index 70032f2c5e..c03d7f6787 100644
--- a/third_party/lua-cjson/lua_cjson.c
+++ b/third_party/lua-cjson/lua_cjson.c
@@ -308,9 +308,9 @@ static void json_append_object(lua_State *l, struct luaL_serializer *cfg,
     while (luaL_next_field(l, cfg) != 0) {
         /* Skip nil values */
         if (lua_isnil(l, -1)) {
-            /* const table, table, key_idx, key, value */
+            /* const table, (table ?), key_idx, key, value */
             lua_pop(l, 1);
-            /* const table, table, key_idx, key */
+            /* const table, (table ?), key_idx, key */
             continue;
         }

@@ -341,10 +341,10 @@ static void json_append_object(lua_State *l, struct luaL_serializer *cfg,
             break;
         }

-        /* const table, table, key_idx, key, value */
+        /* const table, (table ?), key_idx, key, value */
         json_append_data(l, cfg, current_depth, json);
         lua_pop(l, 1);
-        /* const table, table, key_idx, key */
+        /* const table, (table ?), key_idx, key */
     }
     /* const table */

</details>

Reply:
From algorithmic point of view, it is possible. We need to transform the `encode_key_order` array into a hashmap from a key to its ordinal or a next key.

<details>
<summary>Illustration</summary>

```lua
#!/usr/bin/env tarantool

local start_marker = {}
local finish_marker = {}
local next_keys_registry = setmetatable({
    unsorted = {
        [start_marker] = finish_marker,
    },
}, {__mode = 'k'})

local function buf_new()
    return {}
end

local function buf_add(buf, s)
    table.insert(buf, s)
end

local function buf_del(buf)
    table.remove(buf)
end

local function buf_get(buf)
    return table.concat(buf)
end

local function prepare_opts(opts)
    local key_order = opts.encode_key_order
    if key_order == nil or next(key_order) == nil then
        return {
            next_keys = next_keys_registry.unsorted,
        }
    end
    local next_keys = next_keys_registry[key_order]
    if next_keys == nil then
        next_keys = {}
        for i = 1, #key_order do
            local k1 = key_order[i]
            local k2 = key_order[i + 1] or finish_marker
            next_keys[k1] = k2
        end
        next_keys[start_marker] = key_order[1]
        next_keys_registry[key_order] = next_keys
    end
    return {
        next_keys = next_keys,
    }
end

local function encode_kv(buf, key, value)
    buf_add(buf, '"')
    buf_add(buf, key)
    buf_add(buf, '":"')
    buf_add(buf, value)
    buf_add(buf, '"')
end

local function encode(data, opts)
    local prepared_opts = prepare_opts(opts)
    local next_keys = prepared_opts.next_keys

    local buf = buf_new()
    buf_add(buf, '{')

    -- Sorted part.
    local key = next_keys[start_marker]
    while key ~= finish_marker do
        local value = data[key]
        if type(value) ~= 'nil' then
            encode_kv(buf, key, value)
            buf_add(buf, ',')
        end
        key = next_keys[key]
    end

    -- Unsorted part.
    for key, value in pairs(data) do
        if type(next_keys[key]) == 'nil' then
            encode_kv(buf, key, value)
            buf_add(buf, ',')
        end
    end

    buf_del(buf)
    buf_add(buf, '}')

    return buf_get(buf)
end

local data = {a = 'a', b = 'b', c = 'c', d = 'd'}
local opts = {
    encode_key_order = {'a', 'b'},
}
print(encode(data, opts))
```
</details>

From technical point of view, I wouldn't like to review such a code written using the Lua C API.

Reply:
Thank you for working on it. I'm ok with the solution. Sorry, for the delay.